src/protocol/boolean/dumb_bitwise_add_constant.rs

```diff
@@ -102,16 +102,104 @@ where
     Ok(output)
 }
 
+// The Bit Decomposition logic adds a constant value - multiplied by a single sharing of either one or zero.
+// This is a specialized protocol to support that.
+// The win is in cases (like in bit decomposition) where you're adding a constant that's mostly zeroes in binary.
+// In those cases, the carry is very simple, it's just one multiplication, the unknown-bit * the last carry.
+// In those cases where the constant has a 1 bit, you have to do more work,
+// but there are only 2 of those, and one is the very first bit, when there is no other carry! So it's a special case as well.
+pub async fn bitwise_add_constant_maybe<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    a: &[S],
+    b: u128,
+    maybe: &S,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: ArithmeticSecretSharing<F>,
+{
+    let mut output = Vec::with_capacity(a.len() + 1);
+
+    let mut last_carry = S::ZERO;
+    if (b & 1) == 1 {
+        let next_carry = ctx
+            .narrow(&BitOpStep::from(0))
+            .multiply(record_id, &a[0], maybe)
+            .await?;
+        output.push(-next_carry.clone() * F::from(2) + &a[0] + maybe);
+        last_carry = next_carry;
+    } else {
+        output.push(a[0].clone());
+        // In practice, I don't expect to exercize this path, because bit-decomposition
+        // involves (maybe) adding `2^l - PRIME`, which in our case is:
+        // `2^32 - 4_294_967_291 = 5`
+        // ...so the least significant bit will be a `1`.
+        // In fact, for ANY prime number, the least significant bit will ALWAYS be `1`
+        // Since all primes are odd (apart from 2).
```

Comment:
You could probably `assert_eq!(b & 1, 1)` instead.  That would be faster probably.

Reply:
OK. I'll have to update the tests then =)